### PR TITLE
Refactor log server

### PR DESCRIPTION
**Pull request by LiangXugang**
## REFACTOR [ISSUE-#247](https://github.com/caicloud/cyclone/issues/247)

**Changes**
   
  - [ ] refactor: real-time log: websocket --> rest API

**Reviewers**

  - [ ] @supereagle  please review
    
**Checklist**
   
   - [x] Rebased/mergable
   - [x] Tests pass
   - [x] Issue/task done

### CODE REVIEW

File: api/rest/init.go [hunk 1]
```diff
@@ -278,6 +279,16 @@ func registerVersionLogAPIs(ws *restful.WebService) {
...
 		Writes(api.VersionLogCreateResponse{}))
+
+}
+
+func registerVersionLogStreamAPIs(ws *restful.WebService) {
+	ws.Route(ws.GET("/api/v1/projects/{project}/pipelines/{pipeline}/records/{recordID}/logstream").
+		To(getVersionLogStream).
+		Doc("getVersionLogStream get log stream of version").
+		Param(ws.PathParameter("user_id", "identifier of the user").DataType("string")).
+		Param(ws.PathParameter("version_id", "identifier of the version").DataType("string")).
```
Comment (zz-bmj) [inline api/rest/init.go]: not `user_id ` and `version_id `

File: api/rest/log.go [hunk 2]
```diff
@@ -95,3 +118,111 @@ func createVersionLog(request *restful.Request, response *restful.Response) {
...
+		}
+		logstream <- logFragment
+	}
+}
+
+func parseLogFragment(msgFragment string) []byte {
+	var logFragment string
+	if msgFragment != "\r" && msgFragment != "" {
+		if websocket.IsDockerImageOperationLog(msgFragment) {
+			logFragment = msgFragment[6:]
```
Comment (zz-bmj) [inline api/rest/log.go]: What does `6` mean?

Comment (supereagle) [reply api/rest/log.go]: The id of the image layer.

File: api/rest/log.go [hunk 3]
```diff
@@ -95,3 +118,111 @@ func createVersionLog(request *restful.Request, response *restful.Response) {
...
+}
+
+func writerLogStream(ws *socket.Conn, pipelineID string, recordID string, userID string) {
+	pingTicker := time.NewTicker(pingPeriod)
+	defer func() {
+		pingTicker.Stop()
+		ws.Close()
+	}()
+
+	// Read the message from standard input.
```
Comment (supereagle) [inline api/rest/log.go]: This comment?

File: api/rest/log.go [hunk 4]
```diff
@@ -95,3 +118,111 @@ func createVersionLog(request *restful.Request, response *restful.Response) {
...
+			ws.SetWriteDeadline(time.Now().Add(writeWait))
+			if err := ws.WriteMessage(socket.PingMessage, []byte{}); err != nil {
+				log.Error(err.Error())
+				return
+			}
+		}
+	}
+}
+
+func getLogStreamFromKafka(logstream chan []byte, pipelineID string, recordID string, userID string) {
```
Comment (supereagle) [inline api/rest/log.go]: Please add comment for each method.

File: pkg/server/router/logStream.go [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+	pingPeriod = (pongWait * 9) / 10
+)
+
+var upgrader = socket.Upgrader{
+	CheckOrigin:     func(r *http.Request) bool { return true },
+	ReadBufferSize:  1024,
+	WriteBufferSize: 1024,
+}
+
+func getVersionLogStream(request *restful.Request, response *restful.Response) {
```
Comment (supereagle) [inline pkg/server/router/logStream.go]: s/getVersionLogStream/getPipelineRecordLogStream

File: pkg/server/router/logStream.go [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+	projectName := request.PathParameter("project")
+	pipelineName := request.PathParameter("pipeline")
+	recordID := request.PathParameter("recordID")
+
+	pipeline, _, err := rest.FindServiceAndVersion(recordID)
+	if err != nil {
+		response.AddHeader("Content-Type", "text/plain")
```
Comment (supereagle) [inline pkg/server/router/logStream.go]: Refer to other APIs for the handling of error.

File: pkg/server/router/router.go [hunk 7]
```diff
@@ -222,12 +222,22 @@ func (router *router) registerPipelineRecordAPIs(ws *restful.WebService) {
...
+	logdog.Info("Register pipeline records logs APIs")
+
 	// GET /api/v1/projects/{project}/pipelines/{pipeline}/records/{recordId}/logs
 	ws.Route(ws.GET("/projects/{project}/pipelines/{pipeline}/records/{recordId}/logs").To(router.getPipelineRecordLogs).
 		Doc("Get the pipeline record log").
 		Param(ws.PathParameter("project", "name of the project").DataType("string")).
 		Param(ws.PathParameter("pipeline", "name of the pipeline").DataType("string")).
 		Param(ws.PathParameter("recordId", "id of the pipeline record").DataType("string")))
+
+	// GET /api/v1/projects/{project}/pipelines/{pipeline}/records/{recordID}/logstream
```
Comment (supereagle) [inline pkg/server/router/router.go]: s/recordID/recordId/g

File: pkg/server/router/pipelineRecord.go [hunk 8]
```diff
@@ -169,3 +195,115 @@ func (router *router) getPipelineRecordLogs(request *restful.Request, response *
...
+		if errConsume != nil {
+			if errConsume != kafka.ErrNoData {
+				log.Infof("Can't consume %s topic message: %s", sTopic)
+				break
+			} else {
+				continue
+			}
+		}
+
+		//
```
Comment (supereagle) [inline pkg/server/router/pipelineRecord.go]: Remove this line.

**Pull request merged**